dot/parachain/collator-protocol/validator_side.go

```diff
@@ -115,12 +122,323 @@ func (cpvs CollatorProtocolValidatorSide) handleNetworkEvents(event network.Netw
+				if err != nil {
+					// this should never really happen
+					return fmt.Errorf("getting fetched collation from string: %w", err)
+				}
+
+				if fetchedCollation.relayParent == prunedLeaf {
+					delete(cpvs.fetchedCandidates, fetchedCandidateStr)
+				}
+			}
+		}
+
+		// TODO
+		// Remove blocked advertisements that left the view. cpvs.BlockedAdvertisements
+		// Re-trigger previously failed requests again. requestUnBlockedCollations
+		// prune old advertisements
+		// https://github.com/paritytech/polkadot-sdk/blob/aa68ea58f389c2aa4eefab4bf7bc7b787dd56580/polkadot/node/network/collator-protocol/src/validator_side/mod.rs#L1361-L1396
+
+	}
+
+	return nil
+}
+
+func (cpvs *CollatorProtocolValidatorSide) removeOngoing(perRelayParent PerRelayParent) {
+	if perRelayParent.assignment != nil {
+		entry := cpvs.currentAssignments[*perRelayParent.assignment]
+		entry--
+		if entry == 0 {
+			logger.Infof("unassigned from parachain with ID %d", *perRelayParent.assignment)
+		}
+		cpvs.currentAssignments[*perRelayParent.assignment] = entry
+
+	}
+}
+
+func (cpvs *CollatorProtocolValidatorSide) assignIncoming(relayParent common.Hash, perRelayParent *PerRelayParent) error {
+	// TODO: get this instance using relay parent
+	instance, err := cpvs.BlockState.GetRuntime(relayParent)
+	if err != nil {
+		return fmt.Errorf("getting runtime instance: %w", err)
+	}
+
+	validators, err := instance.ParachainHostValidators()
+	if err != nil {
+		return fmt.Errorf("getting validators: %w", err)
+	}
+
+	validatorGroups, err := instance.ParachainHostValidatorGroups()
+	if err != nil {
+		return fmt.Errorf("getting validator groups: %w", err)
+	}
+
+	availabilityCores, err := instance.ParachainHostAvailabilityCores()
+	if err != nil {
+		return fmt.Errorf("getting availability cores: %w", err)
+	}
+
+	validator, validatorIndex := signingKeyAndIndex(validators, cpvs.Keystore)
+	if validator == nil {
+		// return with an error?
+		return nil
+	}
+
+	groupIndex, ok := findValidatorGroup(validatorIndex, *validatorGroups)
+	if !ok {
+		logger.Trace("not a validator")
+		return nil
+	}
+
+	coreIndexNow := validatorGroups.GroupRotationInfo.CoreForGroup(groupIndex, uint8(len(availabilityCores.Types)))
+	coreNow, err := availabilityCores.Types[coreIndexNow.Index].Value()
+	if err != nil {
+		return fmt.Errorf("getting core now: %w", err)
+	}
+
+	paraNow := new(parachaintypes.ParaID)
+	switch coreNow.Index() {
+	case 0:
+		// OccupiedCore
+		occupiedCore, ok := coreNow.(parachaintypes.OccupiedCore)
+		if !ok {
+			return fmt.Errorf("couldn't cast VDT value now to OccupiedCore")
+		}
+		*paraNow = parachaintypes.ParaID(occupiedCore.CandidateDescriptor.ParaID)
+	case 1:
+		// ScheduledCore
+		scheduledCore, ok := coreNow.(parachaintypes.ScheduledCore)
+		if !ok {
+			return fmt.Errorf("couldn't cast VDT value now to scheduledCore")
+		}
+		*paraNow = scheduledCore.ParaID
+	case 2:
+		// Free
+
+	}
+
+	entry := cpvs.currentAssignments[*paraNow]
```

Comment:
in the case `2 (Free)` there is nothing being assigned to `paraNow`, could that be a problem?

Reply:
there is nothing to be done in case of free

Reply:
I see, so in this case it returns None to the `para_now` variable and then on line `523` they check if `para_now` contains a value, if `para_now` is None then they skip the entry update and only updates the `group_assignment`
```
if let Some(para_id) = para_now.as_ref() {}
```

Reply:
changed